### ISSUE TITLE
Main has precedence over module in package.json

### ISSUE THREAD

**Issue by lohfu** [rollup/rollup-plugin-node-resolve#132]
It appears a recent version of this module has changed the resolution priority. From what i can discern `main` suddenly has precedence over `module`. Ie, my build fails because rollup has started pulling in the dependencies' commonjs builds instead of the es builds.

Is this intended?

EDIT:

It appears my usage of `main` combined with `browser` is causing main to have precedence... since `browser` has always had precedence over `module` or `jsnext:main` i used @bfred-it's suggestion from https://github.com/rollup/rollup-plugin-node-resolve/issues/55#issuecomment-310306129.

Version 3.0.2 seems to have updated the resolution algorithm enough to break this way of using `main` and `browser`. Ie. the following no longer works, as rollup will pull in `cjs/index` instead of `es/index`

```
{
  "main": "node/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "browser": {
    "node/index.js": "cjs/index.js"
  }
}
```

EDIT 2:

So this re-raises the question... is there any way of supporting both the `browser` and `module` properties but letting `module` have precedence in ESM aware bundlers, including rollup?

**Comment by M1ras** [comment]
I've run into this issue as well.

Ideally, precedence should be user configurable; `module`, `jsnext`, `main`, and `browser` fields could be assigned a [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) or [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number) value. In the case of a Boolean, precedence is deferred to an internal model. In the case of a Number, 0 comes before 1 and so on. Negative, non-whole, or already used Numbers could throw an [Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error).

**Comment by alex-r-bigelow** [comment]
For testing purposes, trying to use PouchDB is a great way to replicate this problem:

From PouchDB's package.json:
```
"browser": {
    "./lib/index.js": "./lib/index-browser.js",
    "./lib/index.es.js": "./lib/index-browser.es.js"
  }
...
"main": "./lib/index.js",
"module": "./lib/index.es.js",
```
results in this error:
```
[!] Error: 'default' is not exported by node_modules/pouchdb/lib/index-browser.js
```

I've experimented a little with earlier versions (3.0.1, 3.0.0, and 2.1.1) of `rollup-plugin-node-resolve`, and get different (seemingly related*) errors.

* related as far as I can tell... I'm clearly still somewhat of a rollup n00b <rant> in spite of almost a decade of front-end development. I switched from webpack because rollup *looked* more straightforward... but configuring this stuff is so much more hellish than compiling C for different architectures (in rollup's defense, webpack is no better)! And I'm supposed to teach students that are new to programming how to write for the web. In 2018, this is frankly ridiculous. I feel like we're making software development artificially harder for everyone, every time someone comes up with each a cute new idea to "solve" web development. </rant>

**Comment by allex** [comment]
should be fixed in #134

**Comment by alex-r-bigelow** [comment]
Just checked and #134 seems to fix the issue with PouchDB.

**Comment by alex-r-bigelow** [comment]
As a pouchdb-specific workaround, for anyone reaching this via Google, you should use `pouchdb-browser` or `pouchdb-node` (see [custom PouchDB builds](https://pouchdb.com/custom.html)) instead of the full pouchdb dependency when using Rollup. You'll exclude a lot of unnecessary stuff, and avoid the issue described above.

**Comment by VladShcherbin** [comment]
@Rich-Harris hey, this is a pretty big problem because it's currently not possible to use `browser` field properly.

There are already fixes in PR's, can you please look into them?

**Comment by keithamus** [comment]
Im going to close this as many are reporting this is now fixed in `3.0.3`. If you're still experiencing issues, please file a new one or you can comment here and I'll re-open!
